File: robot.py
"""Ported from Robot.java.

AdvantageKit (org.littletonrobotics.junction) has no RobotPy port, so this
version drops it in favor of plain wpilib logging, per team decision: use
commands2.TimedCommandRobot (which schedules CommandScheduler automatically,
same as Java's CommandScheduler.getInstance().run() call in robotPeriodic)
plus wpilib.DataLogManager for on-disk + NetworkTables logging. There is no
AdvantageScope-style replay-log-swap in this version.
"""

from __future__ import annotations

from typing import Optional

import wpilib
from commands2 import Command, TimedCommandRobot

from robotcontainer import RobotContainer


class Robot(TimedCommandRobot):
    def robotInit(self) -> None:
        wpilib.DataLogManager.start()
        wpilib.DataLogManager.logNetworkTables(True)
        wpilib.DriverStation.startDataLog(wpilib.DataLogManager.getLog())

        self._autonomous_command: Optional[Command] = None
        self._robot_container = RobotContainer()

    def disabledInit(self) -> None:
        pass

    def disabledPeriodic(self) -> None:
        pass

    def autonomousInit(self) -> None:
        self._autonomous_command = self._robot_container.get_autonomous_command()

        if self._autonomous_command is not None:
            self._autonomous_command.schedule()

    def autonomousPeriodic(self) -> None:
        pass

    def autonomousExit(self) -> None:
        if self._autonomous_command is not None:
            self._autonomous_command.cancel()

    def teleopInit(self) -> None:
        # This makes sure that the autonomous stops running when teleop starts running.
        if self._autonomous_command is not None:
            self._autonomous_command.cancel()

        # Re-seed pose from vision at teleop start. Covers practice sessions where auto
        # is skipped, and catches any drift that accumulated during auto.
        self._robot_container.initialize_pose()

    def teleopPeriodic(self) -> None:
        pass

    def testInit(self) -> None:
        from commands2 import CommandScheduler

        CommandScheduler.getInstance().cancelAll()
        print("Hello World")

    def testPeriodic(self) -> None:
        pass

    def simulationInit(self) -> None:
        pass

    def simulationPeriodic(self) -> None:
        pass


if __name__ == "__main__":
    wpilib.run(Robot)

File: src/main/java/frc/robot/Robot.java
package frc.robot;

import edu.wpi.first.wpilibj.DataLogManager;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import edu.wpi.first.wpilibj2.command.TimedCommandRobot;

/**
 * Entry point for the teaching-bot proof of concept.
 *
 * <p>{@code TimedCommandRobot} does two things a plain {@code TimedRobot} would leave
 * to you: it calls {@code CommandScheduler.getInstance().run()} every loop
 * automatically (Python's {@code commands2.TimedCommandRobot} does the same), and it
 * still gives you the familiar {@code robotInit()}/{@code autonomousInit()}/
 * {@code teleopInit()}/... callback methods to override.
 *
 * <p>No AdvantageKit, no vision-specific logging (unlike the real competition port's
 * {@code Robot.java}, which extends AdvantageKit's {@code LoggedRobot}) -- just
 * {@link DataLogManager} for on-disk + NetworkTables logging, matching the Python
 * teaching-bot's {@code robot.py} exactly.
 */
public class Robot extends TimedCommandRobot {
    // `Command` (an interface/abstract class) is the TYPE; `m_autonomousCommand` can
    // hold `null` (no autonomous command selected) or any object that implements
    // Command. Python's equivalent used `Optional[Command] = None` as a type hint --
    // Java has no separate "nullable" annotation built into the language the way
    // Python's `Optional[X]` is; ANY non-primitive Java type (anything that isn't
    // `int`/`double`/`boolean`/etc.) can already hold `null`, so `Command` alone is the
    // whole type, and `null` is a value it can take on without any extra syntax.
    private Command m_autonomousCommand;

    // `RobotContainer` is a type WE wrote (see RobotContainer.java) -- Java doesn't
    // distinguish "a class from the standard library" from "a class from this project"
    // in its syntax at all; both are used exactly the same way once imported.
    public RobotContainer m_robotContainer;

    /**
     * This function is run when the robot is first started up and should be used for
     * any initialization code.
     */
    @Override
    public void robotInit() {
        DataLogManager.start();
        DriverStation.startDataLog(DataLogManager.getLog());

        m_robotContainer = new RobotContainer();
    }

    /** This autonomous runs the autonomous command selected by {@link RobotContainer}. */
    @Override
    public void autonomousInit() {
        m_autonomousCommand = m_robotContainer.getAutonomousCommand();

        if (m_autonomousCommand != null) {
            m_autonomousCommand.schedule();
        }
    }

    @Override
    public void autonomousExit() {
        if (m_autonomousCommand != null) {
            m_autonomousCommand.cancel();
        }
    }

    @Override
    public void teleopInit() {
        // This makes sure autonomous stops running when teleop starts. If you want
        // autonomous to continue until interrupted by another command, remove this.
        if (m_autonomousCommand != null) {
            m_autonomousCommand.cancel();
        }
    }

    @Override
    public void testInit() {
        // Cancels all running commands at the start of test mode.
        CommandScheduler.getInstance().cancelAll();
    }
}
